packages/agents-manager/src/components/agent-dock/agent.tsx

```diff
@@ -0,0 +1,220 @@
+
+			// Update the agent's sessionId so future messages use the correct session
+			agentManager.updateSessionId( agentKey, serverSessionId );
+
+			// Only update session if it changed (prevents unnecessary re-renders)
+			if ( chatId !== serverSessionId ) {
+				try {
+					if ( ! serverSessionId ) {
+						// eslint-disable-next-line no-console
+						console.warn( '[AgentDock] Attempted to apply empty session ID' );
+						return;
+					}
+
+					navigate( `/chat/${ serverSessionId }`, { replace: true } );
+				} catch ( error ) {
+					// eslint-disable-next-line no-console
+					console.error( '[AgentDock] Failed to apply session ID:', error );
+					return;
+				}
+			}
+
+			// Track that we've loaded this session (after successful validation)
+			loadedSessionIdRef.current = serverSessionId;
+		},
+		[ agentId, loadMessages, chatId, agentConfig, navigate ]
+	);
+
+	// Conversation loading hook
+	const { loadConversation, isLoading: isLoadingConversation } = useLoadConversation( {
+		apiBaseUrl: API_BASE_URL,
+		authProvider: agentConfig.authProvider,
+		onLoaded,
+	} );
+
+	const abortCurrentRequest = useCallback( async () => {
+		const agentManager = getAgentManager();
+		const agentKey = agentId;
+
+		if ( agentManager.hasAgent( agentKey ) ) {
+			await agentManager.abortCurrentRequest( agentKey );
+		}
+	}, [ agentId ] );
+
+	useEffect( () => {
+		// Navigated to a new chat - reset session
+		if ( ! chatId ) {
+			abortCurrentRequest();
+			lastConversationCache.clear();
+		}
+
+		//Sync store session id
+		setSessionId( chatId );
+	}, [ abortCurrentRequest, chatId, setSessionId ] );
+
+	// Update cache whenever messages change
+	useEffect( () => {
+		if ( ! messages.length || ! chatId ) {
+			return;
+		}
+
+		const agentManager = getAgentManager();
+		const agentKey = agentId;
+
+		if ( ! agentManager.hasAgent( agentKey ) ) {
+			return;
+		}
+
+		// Get Message[] from agentManager and cache it
+		const clientMessages = agentManager.getConversationHistory( agentKey );
+		if ( clientMessages.length ) {
+			const botId = createOdieBotId( agentId );
+			lastConversationCache.set( botId, chatId, clientMessages );
+		}
+	}, [ agentId, messages.length, chatId ] );
+
+	// Load conversation when switching to a session
+	// This handles clicking a conversation from the history list
+	useEffect( () => {
+		if ( ! chatId || isLoadingRef.current ) {
+			return;
+		}
+
+		const agentManager = getAgentManager();
+		// Agent is created by useAgentChat, but might not have messages loaded yet
+		// Check if we need to load messages from server
+		if ( agentManager.hasAgent( agentId ) ) {
+			// Load if this is a different session than what's currently loaded
+			if ( loadedSessionIdRef.current !== chatId ) {
+				isLoadingRef.current = true;
+				const botId = createOdieBotId( agentId );
+				loadConversation( chatId, botId ).finally( () => {
+					isLoadingRef.current = false;
+				} );
+			}
+		}
+	}, [ agentId, loadConversation, chatId ] );
```

Comment:
Speaking from experience, this code will probably suffer from race conditions and memory leaks. Please ignore this if it's just incorrect, but from my perspective, `useAgentChat` should just accept the chatId and the botId and give us all the state we need. I'm surprised we need to do async calls manually inside the compoent.